### PR TITLE
Fix nested client filesync incorrect deduplication

### PR DESCRIPTION
**Pull request by sipsma**
As detailed in #7900, certain obscure corner cases can result in
separate clients on separate hosts (or separate nested exec containers)
from re-using each others' filesync cache refs unexpectedly and, if
files happen to match metadata precisely, end up with incorrect
file contents being used in the sync.

The requirements for this to happen are as obscure as it gets, but it is
impacting our CI and causing flakes, so needs a fix. It is *technically*
possible for users to hit this too, so may as well solve it for them.

Previously part of the cause here was that we used hostname as a
semi-stable host identifier, but nested clients mostly have the same
default container hostname.
* There was much more to it than this, but this is the part we can
  address.

The fix here replaces the hostname with an ID that is persisted in the
client's XDG state dir (i.e. `~/.local/state/dagger/`). It's random when
newly initialized but re-used after that.

That solves this problem for non-nested clients. For nested clients, we
just set this to a random ID in each exec for now.
* It would also be possible to try to persist this in the nested exec,
  but it's not clear yet if that's worth the complication in that case,
  so keeping it simple for now.

### CODE REVIEW

**Comment by sipsma** [comment]
Ah, I actually just remembered that there's a really big tradeoff with relying on hostnames (separate from ones I originally had in mind), which is that it results in CNI network namespaces not being re-usable, which adds likely meaningful performance overhead...

I guess I will just go with the persistent host identifier approach after all.

Also, on my first try here I did get a flake in the `TestPythonProjectLayout` tests still 😢 Was hoping that one might have been the same as #7900 (which is about `TestLegacy`), but either it isn't or there was something wrong about the "unique hostname" approach. Will see what happens after I switch approaches I guess.

**Comment by sipsma** [comment]
Currently letting `TestLegacy` run in a loop for a good while to verify flake is gone. Will do the same for the module tests as a whole. GHA is having an incident right now, but once that's recovered will do re-runs in CI too to check.

**Comment by sipsma** [comment]
Locally:
* `TestLegacy` passed in a loop for an hour
* `TestModule` has passed twice in a row (takes ~30m locally, so gonna keep it looping for a while longer)

In CI:
* Still hitting separate flakes (https://github.com/dagger/dagger/issues/7979 and https://github.com/dagger/dagger/issues/7662)
* But no occurrences of `TestLegacy` flake https://github.com/dagger/dagger/issues/7900

So at least seems to have fixed the expected flake, though sad the others are still going to require their own investigation.

Gonna keep looping in CI a bit more to be extra sure.

**Comment by sipsma** [comment]
Had to do lots of re-runs after other unrelated flakes, but never saw `TestLegacy` flake in CI either, so feel good it's gone.

File: engine/client/stableid.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	// but the worst case is some of the clients using different IDs, which is
+	// just a temporary mild performance deficiency, so not worth more complication
+	tmpFile, err := os.CreateTemp(parentDirPath, StableIDFileName)
+	if err != nil {
+		return "", fmt.Errorf("failed to create stable ID temp file: %w", err)
+	}
+	defer os.Remove(tmpFile.Name())
+	defer tmpFile.Close()
+
+	if err := tmpFile.Chown(uid, gid); err != nil {
```
Comment (jedevc) [inline engine/client/stableid.go:62]: Hello, welcome to windows hell again :tada: `File.Chown` always fails on Windows: https://pkg.go.dev/os#File.Chown :scream:

**Review by jedevc — APPROVED** [review]
Aside from the note that this likely won't work on windows right now (haven't tested to be sure), this looks good to me :tada:

*insane* deep dive, nice job :partying_face:

File: engine/client/stableid.go (hunk 1 above)
Comment (sipsma) [reply engine/client/stableid.go:62]: Oi, I had not encountered that "fun" fact previously, thanks for the pointer! I'll patch this up quick; I actually can drop the chown for now. It was added in an earlier iteration where I was trying to run this code in nested execs too (and thus needed to deal with different uid/gid), but I ended up backing out of that for simplicity.

Comment (sipsma) [reply engine/client/stableid.go]: Fixed in force push

**Pull request merged**